File: android/src/main/kotlin/com/example/wunderkind/WunderkindPlugin.kt
package com.example.wunderkind

import android.content.Context
import android.widget.Toast
import androidx.annotation.NonNull
import java.net.URL

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import co.wunderkind.sdk.Wunderkind
import co.wunderkind.sdk.ProductCategory
import co.wunderkind.sdk.Order
import co.wunderkind.sdk.Invoice
import co.wunderkind.sdk.Customer
import co.wunderkind.sdk.Product
import co.wunderkind.sdk.ScreenType
import co.wunderkind.sdk.other.log.LogLevel
import co.wunderkind.sdk.Currency

/** WunderkindPlugin */
class WunderkindPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private lateinit var context : Context


  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "wunderkind")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
     "initialize" -> handleCall(result) {
        val webId = call.argument<String>("webId")!!
        val isDebugMode = call.argument<Boolean>("isDebugMode")!!
        Wunderkind.getInstance().initialize(context, webId.toLong(), isDebugMode)
      }
      "setIsContextInfoTrackingEnabled" -> handleCall(result) {
        val enabled = call.argument<Boolean>("enabled")!!
        Wunderkind.getInstance().setIsContextInfoTrackingEnabled(enabled)
      }
      "trackScreenView" -> handleCall(result) {
        val urlStr = call.argument<String>("url")!!
        val screenStr = call.argument<String>("screen")!!

        val url = URL(urlStr);
        val screen: ScreenType = ScreenType.valueOf(screenStr.uppercase())
        Wunderkind.getInstance().trackScreenView(url, screen)
      }
      "trackEmptyCart" -> handleCall(result) {
        Wunderkind.getInstance().trackEmptyCart()
      }
      "trackViewItem" -> handleCall(result) {
        val itemId = call.argument<String>("itemId")!!
        val groupId = call.argument<String>("groupId")!!
        Wunderkind.getInstance().trackViewItem(itemId, groupId)
      }
      "trackSelectSku" -> handleCall(result) {
        val groupId = call.argument<String>("groupId")!!
        val feedId = call.argument<String>("feedId")!!
        Wunderkind.getInstance().trackSelectSku(groupId, feedId)
      }
      "trackAddToCart" -> handleCall(result) {
        val itemId = call.argument<String>("itemId")!!
        Wunderkind.getInstance().trackAddToCart(itemId)
      }
      "trackViewCategory" -> handleCall(result) {
        val categoryName = call.argument<String>("categoryName")!!
        val urlStr = call.argument<String>("url")!!
        val ids = call.argument<List<String>>("itemIds")!!

        val url = URL(urlStr);

        val category: ProductCategory = ProductCategory(categoryName, url, ids)
        Wunderkind.getInstance().trackViewCategory(category)
      }
      "trackViewSearch" -> handleCall(result) {
        val searchInput = call.argument<String>("categoryName")!!
        val searchUrlStr = call.argument<String>("url")
        val itemIds = call.argument<List<String>>("itemIds")!!

        val searchUrl = URL(searchUrlStr);
        val searchResults: ProductCategory = ProductCategory(searchInput, searchUrl, itemIds)
        Wunderkind.getInstance().trackViewSearch(searchResults)
      }
      "trackLoggedIn" -> handleCall(result) {
        val phone = call.argument<String>("phone")!!
        val email = call.argument<String>("email")!!

        var phoneStr = phone;

        if(phoneStr.startsWith("+")) {
          phoneStr = phone.drop(1)
        }

        val phoneLong = phoneStr.toLong()
        Wunderkind.getInstance().trackLoggedIn(email, phoneLong)
      }
      "trackLoggedOut" -> handleCall(result) {
        Wunderkind.getInstance().trackLoggedOut()
      }
      "trackTextOptIn" -> handleCall(result) {
        val phone = call.argument<String>("phone")!!

        
        var phoneStr = phone;

        if(phoneStr.startsWith("+")){
          phoneStr = phone.drop(1)
        }

        val phoneLong = phoneStr.toLong()
        Wunderkind.getInstance().trackTextOptIn(phoneLong)
      }
      "trackPurchase" -> handleCall(result) {
        val orderId = call.argument<String>("orderId")!!
        val productsMap = call.argument<List<Map<String, Any>>>("products")!!
        val invoiceMap = call.argument<Map<String, Any>>("invoice")!!
        val paymentMethod = call.argument<String>("paymentMethod")!!
        val customerMap = call.argument<Map<String, Any>>("customer")!!
        val coupons = call.argument<List<String>>("coupons")
        val goal = call.argument<String>("goal")!!

        val currency = invoiceMap["currency"] as String

        val invoice : Invoice = Invoice(
            invoiceMap["amount"] as Double,
            invoiceMap["tax"] as Double,
            invoiceMap["shipping"] as Double,
            invoiceMap["totalDiscount"] as Double?,
            Currency.valueOf(currency.uppercase()),
        );

        var phoneStr = customerMap["phone"] as String;

        if(phoneStr.startsWith("+")) {
          phoneStr = phone.drop(1)
        }

        val phoneLong = phoneStr.toLong()


        val customer : Customer = Customer(
            customerMap["email"] as String,
            phoneLong,
        );

          val products = productsMap.map { map : Map<String, Any> ->
              Product(
                  map["productId"] as String,
                  map["sku"] as String,
                  map["price"] as Double,
                  (map["quantity"] as Int).toLong(),
              )
          }


        val order : Order = Order(
            orderId,
            invoice,
            paymentMethod,
            products,
            customer,
            coupons,
            goal,
        );

        Wunderkind.getInstance().trackPurchase(order)
      }
      "setLogLevel" -> handleCall(result) {
        val levelStr = call.argument<String>("level")!!
        val level: LogLevel = LogLevel.valueOf(levelStr.uppercase())

        Wunderkind.getInstance().setLogLevel(level)
      }

      else -> result.notImplemented()
    }
  }


  private inline fun handleCall(result: Result, block: () -> Unit) {
    try {
      block()
      result.success(true)
    } catch (e: Exception) {
      result.error("WDK Plugin Error", "[WDK Plugin Error] ${e.toString()}", null)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

File: ios/Classes/WunderkindPlugin.swift
import Flutter
import UIKit
import WunderkindKit

public class WunderkindPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "wunderkind", binaryMessenger: registrar.messenger())
    let instance = WunderkindPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      guard let args = call.arguments as? [String : Any] else {return}
      let websiteId = args["webId"] as? String
      let isDebugMode = args["isDebugMode"] as? Bool

      Wunderkind.shared.initialize(websiteId: Int64(websiteId!)!, isDebugMode:isDebugMode ?? true)
      result(nil)
    case "setIsContextInfoTrackingEnabled":
      guard let args = call.arguments as? [String : Any] else {return}
      let enabled = args["enabled"] as? Bool

      Wunderkind.shared.isContextInfoTrackingEnabled = enabled ?? true
      result(nil)
    case "trackScreenView":
      guard let args = call.arguments as? [String : Any] else {return}
      let urlStr = args["url"] as? String
      let screenStr = args["screen"] as? String

      let screenType = WunderkindKit.ScreenType(rawStringValue: screenStr ?? "")

      Wunderkind.shared.trackScreenView(url: URL(string: urlStr ?? "")!, type: screenType ?? .home)
      result(nil)
    case "trackEmptyCart":
      Wunderkind.shared.trackEmptyCart()
      result(nil)
    case "trackViewItem":
      guard let args = call.arguments as? [String : Any] else {return}
      let itemId = args["itemId"] as? String
      let groupId = args["groupId"] as? String

      Wunderkind.shared.trackViewItem(itemId: itemId ?? "", groupId: groupId ?? "") 
      result(nil)
    case "trackSelectSku":
      guard let args = call.arguments as? [String : Any] else {return}
      let groupId = args["groupId"] as? String
      let feedId = args["feedId"] as? String

      Wunderkind.shared.trackSelectSKU(groupId: groupId ?? "", feedId: feedId ?? "")
      result(nil)
    case "trackAddToCart":
      guard let args = call.arguments as? [String : Any] else {return}
      let itemId = args["itemId"] as? String

      Wunderkind.shared.trackAddToCart(itemId: itemId ?? "") 
      result(nil)
    case "trackViewCategory":
      guard let args = call.arguments as? [String : Any] else {return}
      let categoryName = args["categoryName"] as? String
      let urlStr = args["url"] as? String
      let ids = args["itemIds"] as? Array<String>

      var category = WunderkindKit.ProductCategory(title: categoryName ?? "", url: URL(string: urlStr ?? "")!, itemIds: ["test_id"])
      
      Wunderkind.shared.trackViewCategory(category: category)
      result(nil)
    case "trackViewSearch":
      guard let args = call.arguments as? [String : Any] else {return}
      let categoryName = args["categoryName"] as? String
      let urlStr = args["url"] as? String
      let ids = args["itemIds"] as? Array<String>

      var category = WunderkindKit.ProductCategory(title: categoryName ?? "", url: URL(string: urlStr ?? "")!, itemIds: ["test_id"])
      
      Wunderkind.shared.trackViewSearch(category: category) 
      result(nil)
    case "trackLoggedIn":
      guard let args = call.arguments as? [String : Any] else {return}
      let phone = args["phone"] as? String
      let email = args["email"] as? String

      Wunderkind.shared.trackLoggedIn(email: email ?? "", phone: phone ?? "") 
      result(nil)
    case "trackLoggedOut":

      Wunderkind.shared.trackLoggedOut()
    case "trackTextOptIn":
      guard let args = call.arguments as? [String : Any] else {return}
      let phone = args["phone"] as? String

      Wunderkind.shared.trackTextOptIn(phoneNumber: phone ?? "", languageCode: nil) 
      result(nil)
    case "trackPurchase":
      guard let args = call.arguments as? [String : Any] else {return}
      let orderId = args["orderId"] as? String
      guard let productsMap = args["products"] as? Array<[String : Any]> else {return}
      guard let invoiceMap = args["invoice"] as? [String : Any] else {return}
      let paymentMethod = args["paymentMethod"] as? String
      guard let customerMap = args["customer"] as? [String : Any] else {return}
      let coupons = args["coupons"] as? Array<String>
      let goal = args["goal"] as? String
      let currency = invoiceMap["currency"] as? String

      var currencyValue = 32 as Int

      switch (currency) {
        case ("USD"):
            currencyValue = 0
        case ("GBP"):
            currencyValue = 32
        case ("HKD"):
            currencyValue = 34
        case ("EUR"):
            currencyValue = 30
        default:
            currencyValue = 32
        }

      let currencyType = WunderkindKit.Currency(rawValue: currencyValue)
      let invoice = WunderkindKit.Invoice(
          amount: invoiceMap["amount"] != nil ? Decimal(invoiceMap["amount"] as? Double ?? 0.0) : 0.0,
          tax: invoiceMap["tax"] != nil ? Decimal(invoiceMap["tax"] as? Double ?? 0.0) : 0.0,
          shipping: invoiceMap["shipping"] != nil ? Decimal(invoiceMap["shipping"] as? Double ?? 0.0) : 0.0,
          totalDiscount: invoiceMap["totalDiscount"] != nil ? Decimal(invoiceMap["totalDiscount"] as? Double ?? 0.0) : 0.0,
          currency: currencyType ?? .GBP
      )
      let customer = WunderkindKit.Customer(
          email: customerMap["email"] as? String ?? "",
          phone: customerMap["phone"] as? String ?? ""
      )
      var products = [WunderkindKit.Product]()
      for product in productsMap {
        products.append(
          WunderkindKit.Product(
              productId: product["productId"] as? String ?? "",
              skuId: product["sku"] as? String ?? "",
              price: product["price"] as? Decimal ?? 0.0,
              quantity: product["quantity"] as? Int64 ?? 0
          )
        )
      }
      let order = WunderkindKit.Order(
          orderId: orderId ?? "",
          invoice: invoice,
          paymentMethod: paymentMethod ?? "",
          products: products,
          customer: customer,
          coupons: coupons,
          goal: goal ?? ""
      );

      Wunderkind.shared.trackPurchase(order: order)
      result(nil)
    case "setLogLevel":
      guard let args = call.arguments as? [String : Any] else {return}
      let level = args["level"] as? String

      let screenType = WunderkindKit.LogLevel(rawStringValue: level ?? "")
      
      Wunderkind.shared.logLevel = .debug 
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
